src/server/scripts/Kalimdor/DireMaul/instance_dire_maul.cpp

```diff
@@ -59,8 +78,201 @@ class instance_dire_maul : public InstanceMapScript
+                case GO_CRYSTAL_03:
+                    return _crystalGUIDs[2];
+                case GO_CRYSTAL_04:
+                    return _crystalGUIDs[3];
+                case GO_CRYSTAL_05:
+                    return _crystalGUIDs[4];
+                case GO_FORCEFIELD:
+                    return _forcefieldGUID;
+                case NPC_IMMOLTHAR:
+                    return _immoGUID;
+                default:
+                    break;
+            }
+            return ObjectGuid::Empty;
+        }
+
+        void Update(uint32 diff) override
+        {
+            _events.Update(diff);
+
+            while (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_CRYSTAL_CREATURE_STORE:
+                        CrystalCreatureStore();
+                        _events.ScheduleEvent(EVENT_CRYSTAL_CREATURE_CHECK, 3s);
+                        break;
+                    case EVENT_CRYSTAL_CREATURE_CHECK:
+                        CrystalCreatureCheck();
+                        if ((GetBossState(DATA_FORCEFIELD) != DONE))
+                            _events.ScheduleEvent(EVENT_CRYSTAL_CREATURE_CHECK, 1s);
+                        break;
+                    default:
+                         break;
+                }
+            }
+        }
+
+        void CrystalCreatureStore()
+        {
+            for (uint8 i = 0; i < 5; ++i) // we store creatures in a list for all 5 crystals
+            {
+                uint8 creatureCount = 0;
+
+                if (GameObject* crystal = instance->GetGameObject(_crystalGUIDs[i]))
+                {
+                    for (uint8 j = 0; j < 2; ++j)  // once per creature type from CrystalMobs
+                    {
+                        std::list<Creature*> creatureList;
+                        GetCreatureListWithEntryInGrid(creatureList, crystal, CrystalMobs[j], 30.0f);
+                        for (std::list<Creature*>::iterator itr = creatureList.begin(); itr != creatureList.end(); ++itr)
+                        {
+                            _crystalCreatureGUIDs[i][creatureCount] = (*itr)->GetGUID();
+                            ++creatureCount;
+                        }
+                    }
+                }
+            }
+        }
+
+        void CrystalCreatureCheck()
+        {
+            Creature* mob = nullptr;
+            GameObject* go = nullptr;
+
+            for (uint8 i = 0; i < _crystalGUIDs.size(); ++i)
+            {
+                bool _mobAlive = false;
+                go = instance->GetGameObject(_crystalGUIDs[i]);
+                if (!go)
+                    continue;
+
+                if (go->GetGoState() == GO_STATE_READY)
+                {
+                    for (uint8 j = 0; j < _crystalCreatureGUIDs[i].size(); ++j)
+                    {
+                        mob = instance->GetCreature(_crystalCreatureGUIDs[i][j]);
```

Comment:
could "Creature* mob" be declared here or do you need it outside of this loop too ?